aio_cxx: Fix mutual deadlock and resolve test unreliability

NitzanMordhai opened the pull request:
The LibRadosAio.PoolEIOFlag test was unstable, hanging or skipping due to two distinct synchronization problems:

1. Deadlock: The main thread held the shared mutex ('my_lock') while calling thread join. This created a mutual deadlock.
    * Fix: Mutex is unlocked before thread join using RAII scopes.

2. Test Unreliability (Timing Dependency): The test used a fixed iteration count for IO submission and finished too quickly, missing the target error.
    * Fix: The submission loop now runs continuously (time-bounded) until the EIO error is reliably caught, eliminating the timing issue.

Fixes: https://tracker.ceph.com/issues/73946





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

NitzanMordhai committed:
aio_cxx: Fix mutual deadlock in PoolEIOflag test

The LibRadosAio.PoolEIOFlag test was unstable, it was hanging due to:

 - Deadlock: The main thread held the shared mutex ('my_lock') while
   calling thread join. This created a mutual deadlock.
    * Fix: Mutex is unlocked before thread join using RAII scopes.
 - also convert to std::jthread and drop the join

Fixes: https://tracker.ceph.com/issues/73946
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

Matan-B commented on src/test/librados/aio_cxx.cc:
```diff
@@ -2371,25 +2375,37 @@ TEST(LibRadosAio, PoolEIOFlag) {
-      t = new std::thread([&] {
+    // Trigger EIO after 100 ops have been submitted
+    if (i == 100) {
+      t = std::make_unique<std::thread>([&] {
```
Since we already update the code here, how about moving to `std::jthread` to avoid issues like the one here?

Matan-B commented on src/test/librados/aio_cxx.cc:
```diff
@@ -2399,39 +2415,45 @@ TEST(LibRadosAio, PoolEIOFlag) {
-  if (!missed_eio) {
-    my_lock.unlock();
+  {
+    std::unique_lock l(my_lock);
```
nit: We can add a log here "waiting for 60 seconds for ios to complete"

Matan-B reviewed (commented):
Looks good, left one suggestion that might further simplify the fix.
nit, moreover, can you please separate the two changes/fixes in this commit? Should help with smoother review.

NitzanMordhai committed:
aio_cxx: resolve test unreliability

The LibRadosAio.PoolEIOFlag test was unstable, skip due to:
 - Test Unreliability (Timing Dependency): The test used a fixed
   iteration count for IO submission and finished too quickly, missing the
   target error.
    * Fix: The submission loop now runs continuously (time-bounded)
           until the EIO error is reliably caught, eliminating the timing issue.

Fixes: https://tracker.ceph.com/issues/73946
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

NitzanMordhai commented:
@Matan-B thanks! i split the commit and changed your suggestions

sseshasa commented:
Rados Approved.
See: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues74326

SrinivasaBharath commented:
jenkins test api

SrinivasaBharath commented:
jenkins test windows

SrinivasaBharath commented:
jenkins test windows

SrinivasaBharath commented:
@NitzanMordhai -  Kindly fix the issue identified in the failed check and proceed with merging the PR after it has been successfully resolved.

NitzanMordhai commented:
jenkins test windows

NitzanMordhai commented:
jenkins test make check arm64

NitzanMordhai commented:
jenkins test make check arm64
